7 send historical device data via websocket (#76)
hello-nrfcloud/backend: features/HistoricalRequest.feature.md

# Historical data request

> A client can request the history for the device.

## Background

Given I store
`$join(['nrf-historical-', $formatBase($random() * 10000000, 16)])` into
`deviceId`

And a `PCA20035+solar` device with the ID `${deviceId}` is registered with the
fingerprint `92b.d795c7`

## Verify I can query historical device data

Given I connect websocket with fingerprint `92b.d795c7`

And I store
`$toMillis($fromMillis($millis(), '[Y9999]-[M99]-[D99]T[H99]:[m]:00Z'))` into
`ts` and generate a series with decrement of `1` minute each time for `5`
iterations

And I store the converted device messages of device ID `${deviceId}` into
timestream using `${ts}` and decrement the timestamp by `0.5` minutes as this
JSON

> 

Reviewer: Remove this.

```json
[
  {
    "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/gain",

>     "mA": 3.40141

Reviewer: add timestamp property here, as the device would send it using JSONata expressions.

    "mA": 3.75718
  },
  {
    "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/gain",
    "mA": 3.73368
  },
  {
    "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/gain",
    "mA": 3.58041
  },
  {
    "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/gain",
    "mA": 3.79055
  },
  {
    "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/gain",
    "mA": 3.08613
  },
  {
    "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/gain",
    "mA": 3.52402
  },
  {
    "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/gain",
    "mA": 3.04496
  },
  {
    "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/gain",
    "mA": 3.40371
  },
  {
    "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/gain",
    "mA": 3.09479
  }
]
```

And I store `$join(['request-', $formatBase($random() * 10000000, 16)])` into
`requestId`

When I send websocket request as this JSON

```json
{
  "message": "message",
  "payload": {

>     "@id": "${requestId}",

Reviewer: Message must have `@context`, see https://github.com/bifravst/Muninn-backend/issues/47#issuecomment-1592567975

    "type": "lastHour",

>     "ts": ${ts},

Reviewer: ```suggestion

```

Remove.

    "message": "gain",
    "attributes": {
      "avgMA": { "attribute": "mA", "aggregate": "avg" }
    }
  }
}
```

> Then wait for `5` seconds

Reviewer: Remove, use `Soon` keyword.

> Then the historical response should equal to this JSON

Reviewer: ```suggestion
Soon the historical response should equal to this JSON
```

```json
{
  "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/historical-data",
  "@id": "${requestId}",
  "attributes": {
    "avgMA": [
      {
        "mA": 3.579295,
        "ts": ${ts1}
      },
      {
        "mA": 3.657045,
        "ts": ${ts2}
      },
      {
        "mA": 3.43834,
        "ts": ${ts3}
      },
      {
        "mA": 3.28449,
        "ts": ${ts4}
      },
      {
        "mA": 3.24925,
        "ts": ${ts5}
      }
    ]
  },

>   "ts": ${ts}

Reviewer: ```suggestion

```

Should be server time.

# Historical data request

> A client can request the history for the device.

## Background

Given I have the fingerprint for a `PCA20035+solar` device in `fingerprint`

And I store
`$toMillis($fromMillis($millis(), '[Y9999]-[M99]-[D99]T[H99]:[m]:00Z'))` into
`ts`

## Scenario Outline: Device publishes the gain messages to nRF Cloud

Given I store `ts - ${deductMsFromTS}` into `pastTs`

And the device `${fingerprint:deviceId}` publishes this message to the topic
`m/d/${fingerprint:deviceId}/d2c`

```json
{
  "appId": "${appId}",
  "messageType": "DATA",
  "data": "${data}",

>   "ts": ${pastTs}

Reviewer: ```suggestion
  "ts": "$number{pastTs}"
```

}
```

### Examples

| appId | data    | deductMsFromTS |
| ----- | ------- | -------------- |
| SOLAR | 3.40141 | 0              |
| SOLAR | 3.75718 | 30000          |
| SOLAR | 3.73368 | 60000          |
| SOLAR | 3.58041 | 90000          |
| SOLAR | 3.24925 | 120000         |

## Verify I can query gain historical device data

When I connect to the websocket using fingerprint `${fingerprint}`

> And I send websocket request

Reviewer: ```suggestion
And I send this message via the websocket
```

```json
{
  "message": "message",
  "payload": {
    "@context": "https://github.com/hello-nrfcloud/proto/historical-data-request",
    "@id": "420eac59-5ce8-4751-b7d1-217811382095",
    "type": "lastHour",
    "message": "gain",
    "attributes": {
      "avgMA": { "attribute": "mA", "aggregate": "avg" }
    }
  }
}
```

<!-- @retry:tries=5,initialDelay=1000,delayFactor=2 -->

Soon I should receive a message on the websocket that is equal to

```json
{
  "@context": "https://github.com/hello-nrfcloud/proto/historical-data-response",
  "@id": "420eac59-5ce8-4751-b7d1-217811382095",
  "attributes": {
    "avgMA": [
      {
        "mA": 3.40141,

>         "ts": `ts`

Reviewer: ```suggestion
        "ts": "$number{ts}"
```

      },
      {
        "mA": 3.74543,

>         "ts": `ts - 60000`

Reviewer: ```suggestion
        "ts": "$number{ts - 60000}"
```

      },
      {
        "mA": 3.4148300000000003,

>         "ts": `ts - 120000`

Reviewer: ```suggestion
        "ts": "$number{ts - 120000}"
```

      }
    ]
  }
}
```

## Scenario Outline: Device publishes the battery messages to nRF Cloud

Given I store `ts - ${deductMsFromTS}` into `pastTs`

And the device `${fingerprint:deviceId}` publishes this message to the topic
`m/d/${fingerprint:deviceId}/d2c`

```json
{
  "appId": "${appId}",
  "messageType": "DATA",
  "data": "${data}",

>   "ts": ${pastTs}

Reviewer: ```suggestion
  "ts": "$number{pastTs}"
```

        "ts": `ts - 120000`
      }
    ]
  }
}
```

## Verify I will get error message if the request is invalid

When I connect to the websocket using fingerprint `${fingerprint}`

And I send websocket request

```json
{
  "message": "message",
  "payload": {
    "@context": "https://github.com/hello-nrfcloud/proto/historical-data-request",
    "@id": "0c0123ff-f03d-4fc4-934e-0bf1edb7c304",
    "type": "lastHour",
    "message": "unknown",
    "attributes": {
      "avgMA": { "attribute": "mA", "aggregate": "avg" }
    }
  }
}
```

<!-- @retry:tries=5,initialDelay=1000,delayFactor=2 -->

Soon I should receive a message on the websocket that matches

```json
{
  "@context": "https://github.com/hello-nrfcloud/proto/ProblemDetail",
  "@id": "0c0123ff-f03d-4fc4-934e-0bf1edb7c304",
  "type": "https://hello.nrfcloud.com/errors/BadRequest",
  "status": 400,
  "title": "Invalid request"
}
```

## Scenario Outline: Write data into Timestream

Given I store `ts - ${deductMsFromTS}` into `pastTs`

And I write Timestream for the device `${fingerprint:deviceId}` with this
message

```json
{
  "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/location",

>   "lat": ${lat},

Reviewer: ```suggestion
  "lat": "$number{lat}",
```

# Historical data request

> A client can request the history for the device.

## Background

Given I have the fingerprint for a `PCA20035+solar` device in `fingerprint`

And I store
`$toMillis($fromMillis($millis(), '[Y9999]-[M99]-[D99]T[H99]:[m]:00Z'))` into
`ts`

## Scenario Outline: Device publishes the messages to nRF Cloud

> Given I store `ts - ${deductMsFromTS}` into `${storeName}`
> 
> Then the device `${fingerprint:deviceId}` publishes the message with properties
> `${appId}`, `${messageType}`, `${data}`, and `${storeName}` to the topic
> `m/d/${fingerprint:deviceId}/d2c`
> 
> ### Examples
> 
> | appId | messageType | data    | deductMsFromTS | storeName |
> | ----- | ----------- | ------- | -------------- | --------- |
> | SOLAR | DATA        | 3.40141 | 0              | ts1       |
> | SOLAR | DATA        | 3.75718 | 30000          | ts2       |
> | SOLAR | DATA        | 3.73368 | 60000          | ts3       |
> | SOLAR | DATA        | 3.58041 | 90000          | ts4       |
> | SOLAR | DATA        | 3.24925 | 120000         | ts5       |

Reviewer: Should follow established format of devices publishing messages:

```suggestion
Given I store `ts - ${deductMsFromTS}` into `pastTs`

And the device `${fingerprint:deviceId}` publishes this message to the topic
`m/d/${fingerprint:deviceId}/d2c`

```json
{
  "appId": "${appId}",
  "messageType": "DATA",
  "ts": ${pastTs},
  "data": "${data}"
}
```

### Examples

| appId | messageType | data    | deductMsFromTS |
| ----- | ----------- | ------- | -------------- |
| SOLAR | DATA        | 3.40141 | 0              |
| SOLAR | DATA        | 3.75718 | 30000          |
| SOLAR | DATA        | 3.73368 | 60000          |
| SOLAR | DATA        | 3.58041 | 90000          |
| SOLAR | DATA        | 3.24925 | 120000         |

```

## Verify I can query historical device data

When I connect to the websocket using fingerprint `${fingerprint}`

And I send websocket request

```json
{
  "message": "message",
  "payload": {
    "@context": "https://github.com/hello-nrfcloud/proto/historical-data-request",
    "@id": "420eac59-5ce8-4751-b7d1-217811382095",
    "type": "lastHour",
    "message": "gain",
    "attributes": {
      "avgMA": { "attribute": "mA", "aggregate": "avg" }
    }
  }
}
```

<!-- @retry:tries=5,initialDelay=1000,delayFactor=2 -->

Soon I should receive a message on the websocket that matches

```json
{
  "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/historical-data",
  "@id": "420eac59-5ce8-4751-b7d1-217811382095",
  "attributes": {
    "avgMA": [
      {
        "mA": 3.40141,

>         "ts": ${ts1}

Reviewer: No need to use store, these ts can be calculated from the know start `ts` and the expected bin.
